benchmarks/bench_gap_divergence.py

```diff
@@ -0,0 +1,349 @@
+"""
+This benchmark tries to understand what's going on inside the `GapEncoder`,
+to better understand why it is so slow.
+
+The logic is as follows:
+- Tweak the default parameters of the Gap:
+  - Set RNG seed
+  - Set `min_iter` to 5 (`= max_iter`)
+  - Set `max_iter_e_step` to values from 1 to 20
+- Before the end of each loop, save the following:
+  - `W_change`
+  - `score`, from ``self.score(X)``
+  - `W_`
+  - `A_`
+  - `B_`
+- For the visualization, extract from the matrices their:
+  - mean
+  - singular values (`s` of `scipy.linalg.svd`)
+
+Date: June 12th 2023
+Commit: dc77f610e240d2613c99436d01f98db4e4e7922c
+"""
+
+import scipy as sp
+import numpy as np
+import pandas as pd
+import seaborn as sns
+import matplotlib.pyplot as plt
+
+from argparse import ArgumentParser
+from skrub._gap_encoder import (
+    GapEncoder,
+    GapEncoderColumn,
+    _multiplicative_update_h,
+    _multiplicative_update_w,
+    batch_lookup,
+    check_input,
+)
+from joblib import Parallel, delayed
+from sklearn.ensemble import (
+    HistGradientBoostingClassifier,
+    HistGradientBoostingRegressor,
+)
+from sklearn.pipeline import Pipeline
+from sklearn.model_selection import cross_validate
+from skrub import TableVectorizer
+from pathlib import Path
+
+from utils import (
+    monitor,
+    default_parser,
+    find_result,
+    get_classification_datasets,
+    get_regression_datasets,
+)
+
+
+class ModifiedGapEncoderColumn(GapEncoderColumn):
+    def __init__(self, *args, column_name: str = "MISSING COLUMN", **kwargs):
+        super().__init__(*args, **kwargs)
+        self.column_name = column_name
+        self.benchmark_results_: list[dict[str, np.ndarray | float]] = []
+
+    def fit(self, X, y=None):
+        # Copy parameter rho
+        self.rho_ = self.rho
+        # Check if first item has str or np.str_ type
+        assert isinstance(X[0], str), "Input data is not string. "
+        # Make n-grams counts matrix unq_V
+        unq_X, unq_V, lookup = self._init_vars(X)
+        n_batch = (len(X) - 1) // self.batch_size + 1
+        score = self.score(X)
```

Comment:
The score should be computed after the fitting process, otherwise you won't see the impact of hyperparameters.

Reply:
Of course! Silly mistake...